hashintel/hash
# Move Realtime to Redis Streams and adjust its DB schemas

## 🌟 What is the purpose of this PR?



This PR modifies `hash-realtime` to be up to date with `hash-graph`. After we changed our DB layouts, we broke the existing realtime implementation. I've adjusted the setup to use [Redis Streams](https://redis.io/docs/data-types/streams/) and changed the docker compose to use redis-stack (for a graphical Redis interface).

The adjustments are primarily to make sure we're allowing proper pub/sub in Redis and remove as much unnecessary locking as possible. We still have a technically highly-available realtime service that can support automatic failover, but this hasn't fully been utilized/tested yet - we may want to look at consumer groups for Redis Streams in the future.

This work is expected to help external services integrate with realtime data changes from HASH. @indietyp you may be interested in this as well, you can look at the types in `libs/@local/hash-backend-utils/src/pg-tables.ts` to see if you find them useful.



## 🔗 Related links





- [Asana task](https://app.asana.com/0/1202805690238892/1202225566851516/f) _(internal)_ (makes this task redundant)
- [Asana task](https://app.asana.com/0/1204260760205594/1204398654562771/f) _(internal)_ (doesn't complete this)

## 🚫 Blocked by




## 🔍 What does this change?




- Revives realtime using new DB layout (we can't keep the schema in sync automatically)
- Adjusts infra with `turbo prune`
- Clean up environment variables for the services (If you have a `.env.local` please diff and check what has changed!)
- Add realtime to the prod docker-compose

## 🚀 Has this modified a publishable library?




This PR:


## 📜 Does this require a change to the docs?



Nothing outward-facing has changed, existing docs have been updated

## ⚠️ Known issues



## 🐾 Next steps


- Indexing the realtime updates through another loader-service

## 🛡 What tests cover this?




## ❓ How to test this?



1.  Checkout the branch / view the deployment
1.  Start the app (`yarn external-services up` and `yarn dev` in root)
1.  Go to http://localhost:8011/ and observe new update streams in Redis (should be a lot from starting the app)
1.  Sign in to the app and create a type and entities, observe these changes being reflected in their respective streams

## 📹 Demo




https://user-images.githubusercontent.com/6988668/233420579-fe29fcc6-49ef-4609-bcee-cb6063c71891.mp4
